src/script/interpreter.cpp

```diff
@@ -1674,11 +1697,42 @@ static bool VerifyWitnessProgram(const CScriptWitness& witness, int witversion,
         } else {
             return set_error(serror, SCRIPT_ERR_WITNESS_PROGRAM_WRONG_LENGTH);
         }
+    } else if (witversion == 1 && program.size() == WITNESS_V1_TAPROOT_SIZE && !is_p2sh) {
+        // BIP341 Taproot: 32-byte non-P2SH witness v1 program (which encodes a P2C-tweaked pubkey)
+        if (!(flags & SCRIPT_VERIFY_TAPROOT)) return set_success(serror);
+        if (stack.size() == 0) return set_error(serror, SCRIPT_ERR_WITNESS_PROGRAM_WITNESS_EMPTY);
+        if (stack.size() >= 2 && !stack.back().empty() && stack.back()[0] == ANNEX_TAG) {
+            // Drop annex
+            if (flags & SCRIPT_VERIFY_DISCOURAGE_UNKNOWN_ANNEX) return set_error(serror, SCRIPT_ERR_DISCOURAGE_UNKNOWN_ANNEX);
+            SpanPopBack(stack);
+        }
+        if (stack.size() == 1) {
+            // Key path spending (stack size is 1 after removing optional annex)
+            if (!checker.CheckSigSchnorr(stack.front(), program, SigVersion::TAPROOT)) {
+                return set_error(serror, SCRIPT_ERR_TAPROOT_INVALID_SIG);
+            }
+            return set_success(serror);
+        } else {
+            // Script path spending (stack size is >1 after removing optional annex)
+            const valtype& control = SpanPopBack(stack);
+            const valtype& script_bytes = SpanPopBack(stack);
+            scriptPubKey = CScript(script_bytes.begin(), script_bytes.end());
+            if (control.size() < TAPROOT_CONTROL_BASE_SIZE || control.size() > TAPROOT_CONTROL_MAX_SIZE || ((control.size() - TAPROOT_CONTROL_BASE_SIZE) % TAPROOT_CONTROL_NODE_SIZE) != 0) {
+                return set_error(serror, SCRIPT_ERR_TAPROOT_WRONG_CONTROL_SIZE);
+            }
+            if (!VerifyTaprootCommitment(control, program, scriptPubKey)) {
+                return set_error(serror, SCRIPT_ERR_WITNESS_PROGRAM_MISMATCH);
+            }
+            if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_TAPROOT_VERSION) {
```

Comment:
Any reason not to put the cheap policy check before VerifyTaprootCommitment() ?

Reply:
Also perhaps this is checked elsewhere by the logic that sets this SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_TAPROOT_VERSION flag, but don't we need to check the flag version is not currently defined here?

Reply:
Ah I think @theuni commented from an earlier commit, so it wasn't showing that a version is defined.

Reply:
@theuni I think it's preferable to do checks for consensus-invalidity first, so that the error code you get is more consistent. As it isn't inherently more expensive for an attacker to create a policy-invalid spend instead of one that just has an invalid signature say, I don't think it matters for DoS resistance either.

@JeremyRubin I have no idea what you're trying to say.

Reply:
![image](https://user-images.githubusercontent.com/886523/87884608-482e7a80-c9c4-11ea-9e15-4fc36b01c974.png)
@theuni comment shows up on this line, which is later amended to 

![image](https://user-images.githubusercontent.com/886523/87884712-2d103a80-c9c5-11ea-9da4-0fdc12fdd3cb.png)


So I was simply confused from the tagged snippet as to why it did not check the taproot leaf version.

But the reason is because @theuni dropped him comment on an earlier commit, and github didn't tag it as outdated.

Reply:
Ah, I see. That also explains why it may look like these two could easily be swapped @theuni - it's only in a later commit that something is added in between them (and the order matters).

Reply:
I think in theory @theuni's point stands, we could add a cheap check ahead of verifytaprootcommitment that checks and aborts if it's an unknown leaf version, but yes the final state makes more sense as is.